r/R/dataset.R

```diff
@@ -141,6 +141,22 @@ Dataset$create <- function(children, schema) {
 #' @export
 names.Dataset <- function(x) names(x$schema)
 
+#' @export
+dim.Dataset <- function(x) {
+  rows <- sum(purrr::map_dbl(x$files, ~ParquetFileReader$create(.x)$ReadTable()$num_rows))
```

Comment:
IIUC, this is reading an entire table just to count the rows. The row count can be accessed much less expensively by exposing a few more properties of `parquet::arrow::FileReader`

Reply:
Also, a `Dataset` does not necessarily have files (could be a `UnionDataset`) and those files might be in a different format (currently the Ipc format is supported in addition to parquet)

Reply:
Are there specific properties you have in mind?

Reply:
The number of rows is directly available in C++ (it's in the parquet format's metadata). I'd add the following to `r/src/parquet.cpp`:
```c++
// [[arrow::export]]
int64_t parquet___arrow___FileReader__num_rows(
    const std::unique_ptr<parquet::arrow::FileReader>& reader) {
  return reader->parquet_reader()->metadata()->num_rows();
}
```

and expose then expose this in `ParquetFileReader`'s public members in `r/R/parquet.R`

Reply:
Thanks @bkietz! @boshek can you give that a try?

Reply:
The intersection between C++ and R6 here is pushing my limits. Any tips on how to expose an integer [here](https://github.com/apache/arrow/blob/ba22468b88f6aa7793382b4baea78e094f3cfaab/r/R/parquet.R#L459)? I have this:
```     
num_rows = function() {shared_ptr(parquet___arrow___FileReader__num_rows(self))}
```
With this error: `Error in shared_ptr_is_null(xp) : argument "xp" is missing, with no default`

Reply:
I think there's also some `arrow`-specific magic in here too, so it's tricky. You need less magic here though. Omit the `shared_ptr` wrapper, that's for returning a C++ class. 

You can also set this method as an active binding by putting it inside `active()` (you'll need to add it to this class) instead of `public()`.

Reply:
Sorry to be such a pain. Setting as an active binding inside `ParquetFileReader`:
```
active = list(
    num_rows = function() {
      parquet___arrow___FileReader__num_rows(self)
    }
```
Results in 
```
R> pq <- ParquetFileReader$create(pf)
R> pq$num_rows
Error in (function ()  : 
  object 'parquet___arrow___FileReader__num_rows' not found
```
If I define something simple in the active binding it works:

```
active = list(
    num_rows = function() {
      42
    }
  )
```
which results in

```
R> pq <- ParquetFileReader$create(pf)
R> pq$num_rows
[1] 42
```
Any thoughts?

Reply:
Could be a few different things. If you check in your cpp code, I can advise.

Reply:
Checked that in now.

Reply:
Alright. I think the issue is that you need to set `ARROW_R_DEV=true` and install another dependency so that some of the magic gets autogenerated. See https://arrow.apache.org/docs/r/#editing-rcpp-code

Reply:
Yep that works. RTFM 🤦‍♂